drivers/gpio/gpio_psoc6.c

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (c) 2018 Justin Watson
+ * Copyright (c) 2020 ATL Electronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT cypress_psoc6_gpio
+
+#include <errno.h>
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <soc.h>
+#include <drivers/gpio.h>
+
+#include "gpio_utils.h"
+#include "cy_gpio.h"
+#include "cy_sysint.h"
+
+#define LOG_LEVEL CONFIG_GPIO_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(gpio_psoc6);
+
+typedef void (*config_func_t)(const struct device *dev);
+
+struct gpio_psoc6_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	GPIO_PRT_Type *regs;
+	config_func_t config_func;
+};
+
+struct gpio_psoc6_runtime {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	sys_slist_t cb;
+};
+
+#define DEV_CFG(dev) \
+	((const struct gpio_psoc6_config * const)(dev)->config)
+#define DEV_DATA(dev) \
+	((struct gpio_psoc6_runtime * const)(dev)->data)
+
+static int gpio_psoc6_config(const struct device *dev,
+			     gpio_pin_t pin,
+			     gpio_flags_t flags)
+{
+	const struct gpio_psoc6_config * const cfg = DEV_CFG(dev);
+	GPIO_PRT_Type * const port = cfg->regs;
+	uint32_t drv_mode = 0;
+	uint32_t pin_val = 0;
+
+	if (flags & GPIO_SINGLE_ENDED) {
+		drv_mode = (flags & GPIO_OPEN_DRAIN) ?
+			   CY_GPIO_DM_OD_DRIVESLOW_IN_OFF :
+			   CY_GPIO_DM_OD_DRIVESHIGH_IN_OFF;
+	} else if ((flags & GPIO_PULL_UP) && (flags & GPIO_PULL_DOWN)) {
+		drv_mode |= CY_GPIO_DM_PULLUP_DOWN_IN_OFF;
+	} else if (flags & GPIO_PULL_UP) {
+		drv_mode |= CY_GPIO_DM_PULLUP_IN_OFF;
+		pin_val = 1;
+	} else if (flags & GPIO_PULL_DOWN) {
+		drv_mode |= CY_GPIO_DM_PULLDOWN_IN_OFF;
+		pin_val = 0;
+	}
+
+	if (flags & GPIO_INPUT) {
+		/**
+		 * When a Input is defined, the pin initial value is
+		 * fundamental to system proper work!
+		 *
+		 * Because of that, we MUST set pín_val for PULL_UP/DOWN.
+		 */
```

Comment:
It may be better to move the comment to where we set initial pin value. Here it's a bit confusing. Also, we could shorten it. e.g.
```
	} else if (flags & GPIO_PULL_UP) {
		drv_mode |= CY_GPIO_DM_PULLUP_IN_OFF;
		/* Set pin value, necessary to initialize pin as Weak 1 */
		pin_val = 1;
	} else if (flags & GPIO_PULL_DOWN) {
		drv_mode |= CY_GPIO_DM_PULLDOWN_IN_OFF;
		/* Set pin value, necessary to initialize pin as Weak 0 */
		pin_val = 0;
	}
```